File: Bruteforce/codeforces_2063C.py

# Problem link ===>> https://codeforces.com/problemset/problem/2063/C


import sys

def main():
    input = sys.stdin .readline
    for _ in range(int(int().strip())):
        n = int(input().strip())
        
        deg = [0] * n
        adj = [[] for _ in range(n)]
        
        for _ in range(n-1):
            u, v = map(int, input().split())
            u -= 1
            v -= 1
            deg[u] += 1
            deg[v] += 1
            adj[u].append(v)
            adj[v].append(u)
            
        max_result = 0
        sorted_deg = sorted(deg, reverse=True)
        
        for i in range(n):
            node_deg = deg[i]
            neigh_deg = [deg[v] for v in adj[i]]
            neigh_deg.append(node_deg)
            neigh_deg.sort(reverse=True)
            
            max_neigh = -1
            removed = []
            
            for d in neigh_deg:
                if sorted_deg and sorted_deg[-1] == d:
                    sorted_deg.pop()
                    removed.append(d)
                    
            removed.reverse()
            
            if sorted_deg:
                max_neigh = max(max_neigh, sorted_deg[-1])
                
            for v in adj[i]:
                max_neigh = max(max_neigh, deg[v] -1)
                
            for d in removed:
                sorted_deg.append(d)
                
            max_result = max(max_result, node_deg + max_neigh -1)
        print(max_neigh)


if __name__ == "__main__":
    main()

File: Bruteforce/codeforces_2063C.cpp

// Problem link ===>> https://codeforces.com/problemset/problem/2063/C
// submission link ===>> https://codeforces.com/contest/2063/submission/305459632

#include <bits/stdc++.h>
using namespace std;
#define ll long long

void solve()
{
    ll t;
    cin >> t;

    while (t--)
    {
        ll n;
        cin >> n;

        vector<ll> deg(n, 0);
        vector<vector<ll>> adj(n);

        // Input edges and build adjacency list
        for (ll i = 0; i < n - 1; i++)
        {
            ll u, v;
            cin >> u >> v;
            u--, v--; // Convert to 0-based index
            deg[u]++;
            deg[v]++;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        multiset<ll> degreeSet(deg.begin(), deg.end()); // Store degrees in sorted order
        ll max_result = 0;

        // Iterate over each node
        for (ll i = 0; i < n; i++)
        {
            ll node_deg = deg[i];
            vector<ll> neighbor_deg;

            // Collect neighbor degrees
            for (ll v : adj[i])
            {
                neighbor_deg.push_back(deg[v]);
            }

            neighbor_deg.push_back(node_deg);
            sort(neighbor_deg.rbegin(), neighbor_deg.rend()); // Sort in descending order

            ll max_neigh = -1;
            vector<ll> removed;

            // Remove node's degrees temporarily from degreeSet
            for (ll d : neighbor_deg)
            {
                if (!degreeSet.empty() && *degreeSet.rbegin() == d)
                {
                    degreeSet.erase(prev(degreeSet.end()));
                    removed.push_back(d);
                }
            }

            // Get max remaining degree
            if (!degreeSet.empty())
            {
                max_neigh = max(max_neigh, *degreeSet.rbegin());
            }

            // Check max reduced degree of neighbors
            for (ll v : adj[i])
            {
                max_neigh = max(max_neigh, deg[v] - 1);
            }

            // Restore removed elements
            for (ll d : removed)
            {
                degreeSet.insert(d);
            }

            // Update maximum answer
            max_result = max(max_result, node_deg + max_neigh - 1);
        }

        cout << max_result << endl;
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    solve();
    return 0;
}
